srcjs/shinyapp.js

```diff
@@ -714,6 +714,131 @@ var ShinyApp = function() {
     });
   });
 
+  addMessageHandler('shiny-insert-tab', function(message) {
+    var $tabsetPanel = $("#" + message.tabsetPanelId);
+    if ($tabsetPanel.length === 0) {
+      throw 'There is no tabsetPanel with id ' + message.tabsetPanelId;
+    };
+
+    // This is the JS equivalent of the builtItem() R function that is used
```

Comment:
I agree. This was the most intuitive way of coding for me (i.e. try to understand previous code and work with it, rather than refactoring), but in hindsight, that may have been a better bet from the start. @jcheng5, what do you think? Is it worth it?